### DISCUSSION TITLE
What is the minimum amount of pytest's CLI that should be supported by integrating editors?

### DISCUSSION THREAD

**Discussion by brettcannon** [pytest-dev/pytest#8506] [Q&A] [answered]
Slightly odd question, but what I'm wondering is if an editor were to integrate with pytest, what is the minimal amount of pytest's CLI would have to be supported? If we were to tell users that they were expected to use `pytest.ini` and such as much as possible, what would still need to be supported via settings in order for pytest to work?

Assume we know the directory that is open in the editor. Would specifying a different working directory be needed? What about the location of a `pytest.ini` file? Neither of these? Something I'm not thinking of?

Or a slightly different phrasing is what is the minimal amount of custom UI an editor would need to provide to run pytest (assume https://github.com/pytest-dev/pytest/discussions/8458 takes care of discovery/execution UX; this is about running pytest)?

**Comment by nicoddemus** [comment, marked as answer]
Hey @brettcannon 

> what is the minimal amount of pytest's CLI would have to be supported?

As a user, I kind expected to be able to pass any command-line to pytest when executing it inside my editor/IDE. For example, PyCharm has a free form field "Additional arguments" where I can pass any flag I want.

Many flags require command-line arguments which are not expected to be written "permanently" in a the configuration file, but should be used on occasion, for example `--last-failed` or `--stepwise`. The same is true for external plugins.

> Assume we know the directory that is open in the editor. Would specifying a different working directory be needed? What about the location of a pytest.ini file? Neither of these? Something I'm not thinking of?

I'm might be biased, but both PyCharm and PyDev (IIRC about the latter) execute pytest from the directory of the current file, and let pytest itself find the configuration file. I particularly like that default.

> Or a slightly different phrasing is what is the minimal amount of custom UI an editor would need to provide to run pytest

TBH I think an "Additional parameters" is all that's needed, but I might be having tunnel vision because I'm used to having that available to me in PyCharm.

**Reply by brettcannon** [reply]
> As a user, I kind expected to be able to pass any command-line to pytest when executing it inside my editor/IDE.

Do arguments vary between discovery and test execution (I thought they could/did)? So would we need to support a different set depending on what we are executing pytest for (discovery or running)?

> Many flags require command-line arguments which are not expected to be written "permanently" in a the configuration file, but should be used on occasion, for example --last-failed or --stepwise. The same is true for external plugins.

That makes things a bit tricky as then we have to provide some way to let users control whether we prompt them for more flags (and then parse appropriately since we have to get a list of strings to append to list we pass into `pytest.main()` and spaces in shell commands is "interesting" 😉), or some other way to let such ephemeral settings be specified without setting them in a configuration file. But we can't ask every time since that would drive at least me bonkers when I didn't need that feature every time I ran a test. 😄 

> both PyCharm and PyDev (IIRC about the latter) execute pytest from the directory of the current file, and let pytest itself find the configuration file. I particularly like that default.

What happens if you don't specify a file? I.e. if I run all of my tests from the test explorer, where does that set the current working directory? And I assume by "current file" you mean the "current _test_ file" that you have chosen to run a test from and not the specific file you have open which may be the code you are testing and not a test itself? And would this lead to different outcomes compared to discovery since the potential config you find would be different (e.g. `pytest.ini` in the workspace directory compared to one in a testing subdirectory)?

As of right now we let people specify command-line flags in their settings file which we assume people will use to set the directory where their tests are (at least).

What I'm gathering is for test discovery you are suggesting:
```python
args = ["--collect-only", "-pno:terminal"]
args.extend(user_args)
pytest.main(args, plugins=[DiscoveryPlugin()])
```

And then for execution of test(s):
```python
os.chdir(directory_that_DiscoveryPlugin_recorded_for_the_test)
args = ["-pno:terminal"]
args.extend(user_args)
args.append(node_id)  # Drop in the general "run all tests" case.
pytest.main(args, plugins=[ExecutionPlugin()])
```

That seem about right?

**Reply by nicoddemus** [reply]
> Do arguments vary between discovery and test execution (I thought they could/did)? 

The main difference is just the `--collection-only` flag, from the user's perspective. That flag is usually not given given by the user when using the IDE, it is only used internally by the IDE to discover tests.

> So would we need to support a different set depending on what we are executing pytest for (discovery or running)?

I don't think so; I've ever only wanted to have additional flags for running tests, collection/discovery is done by the IDE in the background; I don't think I have ever executed pytest in an IDE using `--collection-only`.

> But we can't ask every time since that would drive at least me bonkers when I didn't need that feature every time I ran a test.

Definitely. What I've seen IDEs use is to have two commands, one that runs the last test command (it being either running a single test, an entire module, or a directory), and another command which opens a configuration dialog that lets you configure the test run before executing. 

> What happens if you don't specify a file? I.e. if I run all of my tests from the test explorer, where does that set the current working directory? 

Depends on the sets of tests being executed. If you are executing the tests from a single test file, set the current working directory to the directory of the test file, if it is a directory containing several test files, set to that directory, and so on.

> And I assume by "current file" you mean the "current test file" that you have chosen to run a test from and not the specific file you have open which may be the code you are testing and not a test itself? 

Ahh yes sorry.

But this reminds me of an awesome feature that I often miss, where the editor is smart enough to detect the test file associated with a module/package, and have a shortcut to execute those tests. 

Currently there's no explicit association between a module and a test file, so for something like this to work it would have to assume some convention, for example, if I have `x.py` open, that shortcut would execute the test module `test_x.py` in the workspace (if it exists).

> And would this lead to different outcomes compared to discovery since the potential config you find would be different (e.g. pytest.ini in the workspace directory compared to one in a testing subdirectory)?

That's not supported by pytest, we only support a single configuration file, so this shouldn't be a problem. pytest is designed so that the current working directory doesn't really matter when you execute a test: we search for the config file based on the tests given in the command-line, not on the current working directory (except when you don't pass any test file or directory, in which case we use the current working directory to find the config file).

> That seem about right?

It does, however you probably don't need two separate plugins, you can have just one.

**Reply by brettcannon** [reply]
> Depends on the sets of tests being executed. If you are executing the tests from a single test file, set the current working directory to the directory of the test file, if it is a directory containing several test files, set to that directory, and so on.

So I'm thinking of the bare `pytest` scenario where pytest traverses down, collects everything, and starts running tests. Now let's say that the tests are all in a `tests/` directory. Would that mean you expect the working directory to be where you executed `pytest` from or `tests/`? It seems like the latter, but I want to make sure I'm understanding clearly.

> But this reminds me of an awesome feature that I often miss, where the editor is smart enough to detect the test file associated with a module/package, and have a shortcut to execute those tests.

You are definitely not the first to ask for this. 😅

> we search for the config file based on the tests given in the command-line, not on the current working directory (except when you don't pass any test file or directory, in which case we use the current working directory to find the config file).

One interesting thing I just realized is how this will muck with relative file paths in tests (e.g. `open()` to load some golden file). Might still need to provide a way for people to override where they want to anchor their execution for the working directory but default to the semantics you're suggesting.

**Reply by nicoddemus** [reply]
> It seems like the latter, but I want to make sure I'm understanding clearly.

I agree. However this is a convention decided by the IDE. I would also understand/accept if the IDE decides to always run the tests using the directory containing `pytest.ini` as current working directory, for example.

> You are definitely not the first to ask for this. 😅

Heh!

> One interesting thing I just realized is how this will muck with relative file paths in tests (e.g. open() to load some golden file).

Definitely. While the IDE should have an option to control this, I think tests that depend on the current working directory to work are an anti-pattern, users should be getting relative paths based on `__file__`, or providing those golden file paths through a fixture (which would also use `__file__` to locate it). That's just a comment of course, the IDE can't prevent those anti-patterns so it should at least have clear rules to which current working directory it will use to run tests.

**Reply by brettcannon** [reply]
>  I would also understand/accept if the IDE decides to always run the tests using the directory containing pytest.ini as current working directory, for example.

Is there a way that a plug-in gets told where the config file lives? I quick check of the docs didn't yield anything obvious for me.

> While the IDE should have an option to control this, I think tests that depend on the current working directory to work are an anti-pattern

Totally agree! I just know I have answered my fair share of Stack Overflow questions caused by this sort of thing. 😄

**Reply by RonnyPfannschmidt** [reply]
@brettcannon there is  `config.inipath/inifile` for the ini config, and there is the `tool.pytest` section for `pyproject.toml`

**Reply by brettcannon** [reply]
@RonnyPfannschmidt Thanks! Found https://docs.pytest.org/en/stable/reference.html#pytest.config.Config.inipath.
